File: LC/Python/Medium/2107.py
class Solution:
    def shareCandies(self, candies: List[int], k: int) -> int:
        ans = 0
        seen = Counter(candies)
        for window_end in range(len(candies)):
            seen[candies[window_end]] -= 1 
            
            if seen[candies[window_end]] == 0: 
                del seen[candies[window_end]]
            
            if window_end >= k: 
                seen[candies[window_end - k]] += 1 
            
            if window_end >= k - 1: 
                ans = max(ans, len(seen))

        return ans 

File: LC/C++/Medium/2107.cpp
class Solution {
public:
    int shareCandies(vector<int>& candies, int k) {
        int ans{0};
        std::unordered_map<int, int> window{};
        for (auto x : candies) window[x]++;

        for (int i = 0; i < candies.size(); ++i) {
            if (--window[candies[i]] == 0) {
                window.erase(candies[i]);
            }

            if (i >= k) {
                window[candies[i - k]]++;
            }

            if (i >= k - 1) {
                ans = std::max(ans, (int)window.size());
            }
        }

        return ans;
    }
};